src/sagemaker_containers/functions.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License'). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import inspect
+
+import six
+
+import sagemaker_containers as smc
+
+
+def matching_args(fn, dictionary):  # type: (function, collections.Mapping) -> dict
+    """Given a function fn and a dict dictionary, returns the function arguments that match the dict keys.
+
+    Example:
+
+        def train(channel_dirs, model_dir): pass
+
+        dictionary = {'channel_dirs': {}, 'model_dir': '/opt/ml/model', 'other_args': None}
+
+        args = smc.functions.matching_args(train, dictionary) # {'channel_dirs': {}, 'model_dir': '/opt/ml/model'}
+
+        train(**args)
+    Args:
+        fn (function): a function
+        dictionary (dict): the dictionary with the keys
+
+    Returns:
+        (dict) a dictionary with only matching arguments.
+    """
+    args, _, kwargs = signature(fn)
+
+    if kwargs:
+        return dictionary
+
+    return smc.collections.split_by_criteria(dictionary, set(args))[0]
+
+
+def signature(fn):  # type: (function) -> ([], [], [])
+    """Given a function fn, returns the function args, vargs and kwargs
+
+    Args:
+        fn (function): a function
+
+    Returns:
+        ([], [], []): a tuple containing the function args, vargs and kwargs.
```

Comment:
I think we should return NamedTuples instead of regular tuples for all public functions unless we have a strong reason not to, since they still support the functionality of tuples, but gives the user a way to write more readable code if they want by referencing field names instead of tuple indices. An example is in the current matching_args implementation, which currently has:

smc.collections.split_by_criteria(dictionary, set(args))[0]

Which can become:

smc.collections.split_by_criteria(dictionary, set(args)).included (or whatever name we come up with for the field.)

You can see examples of this from the Python standard library, such as inspect.getargspec returning a ArgSpec named tuple.

If we agree on this, we should go back and apply this to the other places in our code where we return tuples (or revisit whether returning tuples at all is the best solution).

Reply:
@winstonaws great point. Will do that